cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,311 @@
+  private final DatasetFramework datasetFramework;
+  private final Store store;
+  private final ProgramLifecycleService lifecycleService;
+  private final PropertiesResolver propertiesResolver;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final CConfiguration cConf;
+  private ScheduleTaskRunner taskRunner;
+  private ListeningExecutorService taskExecutorService;
+
+  @Inject
+  ConstraintCheckerService(Store store,
+                           ProgramLifecycleService lifecycleService, PropertiesResolver propertiesResolver,
+                           NamespaceQueryAdmin namespaceQueryAdmin,
+                           CConfiguration cConf,
+                           DatasetFramework datasetFramework,
+                           TransactionSystemClient txClient) {
+    this.store = store;
+    this.lifecycleService = lifecycleService;
+    this.propertiesResolver = propertiesResolver;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.cConf = cConf;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(datasetFramework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+    this.datasetFramework = datasetFramework;
+  }
+
+  @Override
+  protected void doStart() {
+    if (!isRunning()) {
+      return;
+    }
+    LOG.info("Starting ConstraintCheckerService.");
+    taskExecutorService = MoreExecutors.listeningDecorator(
+      Executors.newCachedThreadPool(new ThreadFactoryBuilder().setNameFormat("constraint-checker-task").build()));
+    taskRunner = new ScheduleTaskRunner(store, lifecycleService, propertiesResolver,
+                                        taskExecutorService, namespaceQueryAdmin, cConf);
+
+    for (int partition = 0; partition < getNumPartitions(); partition++) {
+      taskExecutorService.submit(new ConstraintCheckerThread(partition));
+    }
+    LOG.info("Started ConstraintCheckerService.");
+  }
+
+  @Override
+  protected void doStop() {
+    LOG.info("Stopping ConstraintCheckerService.");
+    try {
+      if (!taskExecutorService.awaitTermination(5, TimeUnit.SECONDS)) {
+        taskExecutorService.shutdownNow();
+      }
+    } catch (InterruptedException ie) {
+      Thread.currentThread().interrupt();
+    }
+    LOG.info("Stopped ConstraintCheckerService.");
+  }
+
+  private int getNumPartitions() {
+    try {
+      return Transactions.execute(transactional, new TxCallable<Integer>() {
+        @Override
+        public Integer call(DatasetContext context) throws Exception {
+          return Schedulers.getJobQueue(context, datasetFramework).getNumPartitions();
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private class ConstraintCheckerThread extends Thread {
```

Comment:
Since you are now using `Executor`, this should be `Runnable` instead of `Thread`.